nucypher/blockchain/economics.py

```diff
@@ -14,60 +14,166 @@
+                 small_stake_multiplier: Decimal = __default_small_stake_multiplier
+                 ):
         """
-        Calculate parameters to use in token issuer from high-level
-        human-understandable parameters
-
-        :param initial_supply: coins at t=0,
-        :param initial_inflation: inflation in day 1 expressed in units of year**-1
-            e.g. 1/365 of all coins in day 1 is 1 year**-1
-        :param T_half: time for inflation halving in years
-        :param T_sat: "saturation" time - if staking is longer than T_sat, the
-            reward doesn't get higher
-        :param small_staker_multiplier: what fraction of max reward rate do we
-            pay to those who are about to unlock coins
+        :param initial_supply: Tokens at t=0
+        :param initial_inflation: Inflation on day 1 expressed in units of year**-1
+        :param halving_delay: Time for inflation halving in years
+        :param reward_saturation: "saturation" time - if staking is longer than T_sat, the reward doesn't get any higher
+        :param small_stake_multiplier: Fraction of maximum reward rate paid to those who are about to unlock tokens
         """
 
-        # Saving input in case we'll need it
-        self.input_params = dict(
-                initial_supply=initial_supply,
-                initial_inflation=initial_inflation,
-                T_half=T_half,
-                T_sat=T_sat,
-                small_staker_multiplier=small_staker_multiplier)
+        self.initial_supply = initial_supply
+        self.initial_inflation = initial_inflation
+        self.token_halving = halving_delay
+        self.token_saturation = reward_saturation
+        self.small_stake_multiplier = small_stake_multiplier
+        self.log_scale = Decimal(log(2))
+
+        #
+        # Calculate
+        #
+
+        # ERC20 Token parameter
+        self.__total_supply = int(initial_supply * (1 + initial_inflation * halving_delay / self.log_scale))
+
+        # Remaining / Reward Supply - Escrow Parameter
+        self.__reward_supply = int(self.__total_supply - self.initial_supply)
+
+        # k2 - Escrow parameter
+        self.__staking_coefficient = 365 ** 2 * reward_saturation * halving_delay / self.log_scale / (1 - small_stake_multiplier)
+
+        # k1 - Escrow parameter
+        self.__locked_periods_coefficient = 365 * reward_saturation * small_stake_multiplier / (1 - small_stake_multiplier)
+
+        # Maximum locked periods- Escrow parameter
+        self.__awarded_periods = reward_saturation * 365
+
+    @property
+    def k1(self) -> int:
+        return int(self.__locked_periods_coefficient)
+
+    @property
+    def k2(self) -> int:
+        return int(self.__staking_coefficient)
+
+    @property
+    def locked_periods_coefficient(self) -> int:
+        return self.__locked_periods_coefficient
+
+    @property
+    def staking_coefficient(self) -> int:
+        return self.__staking_coefficient
```

Comment:
What's the point of wrapping these as `@property`, why not setting `self.staking_coefficient` directly in `__init__`?

Reply:
Mainly this is an attempt to protect these attributes from being mutated later.

Reply:
:heavy_check_mark: